mesonbuild/dependencies/dev.py

```diff
@@ -700,6 +702,107 @@ def __init__(self, environment: 'Environment', kwargs: JNISystemDependencyKW):
 packages['jdk'] = JDKSystemDependency
 
 
+class DiaSDKSystemDependency(SystemDependency):
+
+    def _try_path(self, diadir: str, cpu: str) -> bool:
+        if not os.path.isdir(diadir):
+            return False
+
+        include = os.path.join(diadir, 'include')
+        if not os.path.isdir(include):
+            mlog.error('DIA SDK is missing include directory:', include)
+            return False
+
+        lib = os.path.join(diadir, 'lib', cpu, 'diaguids.lib')
+        if not os.path.exists(lib):
+            mlog.error('DIA SDK is missing library:', lib)
+            return False
+
+        bindir = os.path.join(diadir, 'bin', cpu)
+        if not os.path.exists(bindir):
+            mlog.error(f'Directory {bindir} not found')
+            return False
+
+        found = glob.glob(os.path.join(bindir, 'msdia*.dll'))
+        if not found:
+            mlog.error("Can't find msdia*.dll in " + bindir)
+            return False
+        if len(found) > 1:
+            mlog.error('Multiple msdia*.dll files found in ' + bindir)
+            return False
+        self.dll = found[0]
+
+        # What to do about version?
+        # At runtime, the DIA interface version used is controlled by GUID passed to CoCreateInstance.
+        # Newer DLLs contain backwards support for all older GUID.
+        # DIA SDK include files only provide GUIDS for the latest version.
+        # Finding out version of current SDK is tricky. The only place where the actual SDK
+        # version is stored, seems to be the Version field in msdia*.dll resources.
+        # To get the just the major version, we could read it from msdiaNNN.dll file name:
+        self.version = 'unknown'
```

Comment:
Given the comment about the version here, I think if there are more than one `msdiaNNN.dll` we should just return the highest version number if they haven't requested a specific version, since it should be backwards compatible? If they need a specific version, we need to do something more.

On the more version information... that's tricky. Since the SDK headers only contain the GUIDS for the latest version I only see a couple of choices for us:
  1. We only handle major versions, and just use the `NNN` in the `.dll` name
  2. We only handle `>` and `>=` for the version, and use the header GUID.
  3. We maintain our own internal list of version mappings. I like this the least of our choices.

Reply:
My impression is that with DIA SDK, the SDK version is mostly an internal thing that should be opaque to library user. I couldn't even find it listed anywhere publicly. The only place where you get to see anything with version (but in an exceptionaly bad way) is the major version in DLL name. So, out of these, I think #1 should be enough.

Reply:
Sounds good to me.

Reply:
@dcbaker: Please see https://github.com/mesonbuild/meson/pull/13638/commits/10f9fa10659ab4341c4683c7a7a24bf41915ff9e .

I don't know how to test this with meson test suite, eg. how to check for particular error after `dependency('diasdk', version: '> 150')` for example.

Reply:
BTW, as I understand things, there should never be more than single msdiaNNN.dll in the `DIA SDK/bin/<arch>` directory (unless the user manually copies them there). This would be a completely unexpected situation, that I don't see reason to handle, unless we find out microsoft started shipping the SDK like that. I just didn't want to leave the hypothetical situation unchecked in code, and give us warning if it ever happens.

Reply:
Done